iree/compiler/Dialect/Stream/Transforms/ScheduleAllocation.cpp

```diff
@@ -0,0 +1,1440 @@
+  auto sourceRange = scope.lookupResourceRange(asyncOp.source());
+  auto sourceOffset =
+      scope.add(asyncOp.getLoc(), sourceRange.offset, asyncOp.source_offset());
+  auto targetRange = scope.lookupResourceRange(asyncOp.result());
+  builder.create<IREE::Stream::CmdCopyOp>(
+      asyncOp.getLoc(), sourceRange.resource, sourceRange.resourceSize,
+      sourceOffset, targetRange.resource, targetRange.resourceSize,
+      targetRange.offset, asyncOp.result_size());
+  asyncOp.erase();
+  return success();
+}
+
+static LogicalResult applyAsyncFillOp(IREE::Stream::AsyncFillOp asyncOp,
+                                      AllocationScope &scope,
+                                      OpBuilder builder) {
+  auto targetRange = scope.lookupResourceRange(asyncOp.result());
+  auto targetOffset =
+      scope.add(asyncOp.getLoc(), targetRange.offset, asyncOp.target_offset());
+  builder.create<IREE::Stream::CmdFillOp>(
+      asyncOp.getLoc(), targetRange.resource, targetRange.resourceSize,
+      targetOffset, asyncOp.target_length(), asyncOp.value());
+  asyncOp.erase();
+  return success();
+}
+
+static LogicalResult applyAsyncUpdateOp(IREE::Stream::AsyncUpdateOp asyncOp,
+                                        AllocationScope &scope,
+                                        OpBuilder builder) {
+  // TODO(benvanik): place the allocation instead.
+  auto sourceRange = scope.lookupResourceRange(asyncOp.update());
+  auto sourceOffset = sourceRange.offset;
+  auto targetRange = scope.lookupResourceRange(asyncOp.result());
+  auto targetOffset =
+      scope.add(asyncOp.getLoc(), targetRange.offset, asyncOp.target_offset());
+  builder.create<IREE::Stream::CmdCopyOp>(
+      asyncOp.getLoc(), sourceRange.resource, sourceRange.resourceSize,
+      sourceOffset, targetRange.resource, targetRange.resourceSize,
+      targetOffset, asyncOp.update_size());
+  asyncOp.erase();
+  return success();
+}
+
+static LogicalResult applyAsyncCopyOp(IREE::Stream::AsyncCopyOp asyncOp,
+                                      AllocationScope &scope,
+                                      OpBuilder builder) {
+  auto sourceRange = scope.lookupResourceRange(asyncOp.source());
+  auto sourceOffset =
+      scope.add(asyncOp.getLoc(), sourceRange.offset, asyncOp.source_offset());
+  auto targetRange = scope.lookupResourceRange(asyncOp.result());
+  auto targetOffset =
+      scope.add(asyncOp.getLoc(), targetRange.offset, asyncOp.target_offset());
+  builder.create<IREE::Stream::CmdCopyOp>(
+      asyncOp.getLoc(), sourceRange.resource, sourceRange.resourceSize,
+      sourceOffset, targetRange.resource, targetRange.resourceSize,
+      targetOffset, asyncOp.length());
+  asyncOp.erase();
+  return success();
+}
+
+static LogicalResult applyAsyncTransferOp(IREE::Stream::AsyncTransferOp asyncOp,
+                                          AllocationScope &scope,
+                                          OpBuilder builder) {
```

Comment:
I like the structure of this function - it reads easily and the branches (if transferIn, if transferOut) are inserted cleanly :)